### DISCUSSION TITLE
Where is the `.wasm`?

### DISCUSSION THREAD

**Discussion by jymchng** [x1unix/go-playground#429] [Q&A] [answered]
Hi, thank you for your hard work on this playground.

I am interested to build a Go playground as well with WASM, I see that you guys are using WASM execute arbitrary Go 1.22 codes, I am curious to know where is the `.wasm` binary for this.

Thank you.

**Comment by x1unix** [comment, marked as answer]
Hello @jymchng 

WASM binary is built on a server and then served to a user for execution. \
Under the hood, API serves a cache or otherwise - runs `go build` to build a WASM program.

**See:**

* Backend:
  - https://github.com/x1unix/go-playground/blob/master/internal/server/handler_v2.go#L200 - HTTP handler
  - https://github.com/x1unix/go-playground/blob/master/internal/builder/compiler.go#L85 - Builder business logic
* Frontend:
  - https://github.com/x1unix/go-playground/blob/master/web/src/store/dispatchers/build/dispatch.ts#L127

P.S - Previously we had a in-browser Go interpreter based on [Yaegi](https://github.com/traefik/yaegi) but it was removed as project is not actively maintained and doesn't support modern Go features like generics.

In future we plan to move Go compiler completely into a browser but this is a complex task.

**Reply by jymchng** [reply]
> Hello @jymchng
> 
> WASM binary is built on a server and then served to a user for execution. Under the hood, API serves a cache or otherwise - runs `go build` to build a WASM program.
> 
> **See:**
> 
> * Backend:
>   
>   * https://github.com/x1unix/go-playground/blob/master/internal/server/handler_v2.go#L200 - HTTP handler
>   * https://github.com/x1unix/go-playground/blob/master/internal/builder/compiler.go#L85 - Builder business logic
> * Frontend:
>   
>   * https://github.com/x1unix/go-playground/blob/master/web/src/store/dispatchers/build/dispatch.ts#L127
> 
> P.S - Previously we had a in-browser Go interpreter based on [Yaegi](https://github.com/traefik/yaegi) but it was removed as project is not actively maintained and doesn't support modern Go features like generics.
> 
> In future we plan to move Go compiler completely into a browser but this is a complex task.

Yeah, moving the entire Go compiler into the browser is very difficult although I don't fully understand it. I had asked TinyGo this question before and they replied it is the CGo part of the compiler that is blocking them.

**Reply by x1unix** [reply]
> I had asked TinyGo this question before and they replied it is the CGo part of the compiler that is blocking them.

If you're doing WASM builds - you can throw away cgo part. Probably writing own compiler based on original Go compiler will be more simple approach, as `go build` calls a lot of commands under the hood and getting `exec.Cmd` working in browser is complicated.

If you're looking for an easy path and don't need generics - Yaegi might be suitable for you.

**Comment by jymchng** [comment]
Thank you for your reply. @x1unix 

Essentially, what you're saying is the code execution actually uses the actual Go compiler to compile the source files on the playground into a `.wasm` binary and execute the `.wasm` binary. There's isn't an in-browser `.wasm` binary that is responsible for compiling and executing the `.go` source files on the playground.

Is my interpretation accurate?

**Reply by x1unix** [reply]
Yes, correct. All compilation is happening on server.
